Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp

```diff
@@ -343,6 +344,16 @@ void webKitMediaSrcEmitStreams(WebKitMediaSrc* source, const Vector<RefPtr<Media
     gst_element_post_message(GST_ELEMENT(source), gst_message_new_stream_collection(GST_OBJECT(source), source->priv->collection.get()));
 
     for (const RefPtr<Stream>& stream: source->priv->streams.values()) {
+        // Block data flow until pad is exposed.
+        gulong blockId = gst_pad_add_probe(
+            GST_PAD(stream->pad.get()), GST_PAD_PROBE_TYPE_BLOCK_DOWNSTREAM,
+            [](GstPad*, GstPadProbeInfo*, gpointer) -> GstPadProbeReturn {
+                return GST_PAD_PROBE_OK;
+            }, nullptr, nullptr);
+        auto blockCleanup = WTF::makeScopeExit([&] {
+            gst_pad_remove_probe(GST_PAD(stream->pad.get()), blockId);
+        });
```

Comment:
Documentation about `GST_PAD_PROBE_REMOVE` is incomplete but I would check if it passes the data and if it does, you don't need the scoped exit.

Reply:
Manually removing the probe is the suggested way to remove a **blocking** probe:
https://gstreamer.freedesktop.org/documentation/additional/design/probes.html?gi-language=c#blocking-probes

There's also the possibility of returning GST_PAD_PROBE_REMOVE, but I'm not sure if the data would be dropped or returned in that case, so better stick to `gst_pad_remove_probe()` in this case. The makeScopeExit ensures that the probe is eventually removed, either if data has ever traversed the pad or not. That's the main difference: gst_pad_remove_probe() would work even if there's no data, and that's an advantage that justifies keeping this code block as is. Do you agree?

Reply:
Well:

> The blocking probe is removed with gst_pad_remove_probe() or when the probe callback return GST_PAD_PROBE_REMOVE. In both cases, and if this was the last blocking probe on the pad, the pad is unblocked and dataflow can continue.

Reply:
I'd also recommend to remove the scope exit and add it to the end of the scope just after line 369 because the code is not error prone with early returns to forget this.

Reply:
And also, even when the probe is blocking, to serve it's purpose, you should check inside it if pad is active and wait there until it is. Otherwise the blocking probe will block, execute and unblock, even with the pad not being active yet, right?

Reply:
It looks like I a mistaken on the blocking stuff, it should work but still, the scope exit is not needed.